### ISSUE TITLE
Create dialog with Bootstrap's feature instead of jQuery-ui

### ISSUE THREAD

**Issue by kz0817** [project-hatohol/hatohol#164]
Now we are using jQuery-ui to make a dialogbox.
However, our main CSS: Bootstrap has a similar feature such as modal.
So we should use Bootstrap's one and delete jQuery-ui to make file composition of the repository simple.

Ref.
http://getbootstrap.com/javascript/

**Comment by ashie** [comment]
以前試しにやろうとした残骸を experimental/bootstrap-modal に入れておきます。

**Comment by cosmo0920** [comment]
あざっす！

**Comment by cosmo0920** [comment]
メモ：
https://github.com/project-hatohol/hatohol/compare/experimental/feature-test-casperjs
[Casper.js 1.1 beta3](http://docs.casperjs.org/en/latest/) の調査も兼ねて、featureテストがどんな感じになるかを実験したもの。
モーダルダイアログをBootstrap由来のものに置き換えるなら上記ブランチのようなUIに関するテストが必要そう。

@Mnakagawa さんが作っていた https://github.com/project-hatohol/web-gui-test を動かす事ができれば足りない分を補ってテストを行いたい。

**Comment by cosmo0920** [comment]
@Mnakagawa さん、 https://github.com/project-hatohol/web-gui-test の動かし方が分からないのですがどこかに書いてありますか？

**Comment by Mnakagawa** [comment]
@cosmo0920 さん
以下コマンドを実行し、selenium webdriverをインストールしてください。
#pip install selenium
ライブラリがインストールされるので、後は通常のスクリプト同様に.py形式の
テストを実行してください。ブラウザ（firefox or Google chrome）が立ち上がり
自動的にテストを行ってくれます。
現段階ではログインログアウト、また登録が必要なもの（zabbix,nagios,redmineなど）の
登録テストが存在していますが、３月から増やしていく予定です。

新規でテストを作成する場合は、firefoxのプラグインのselenium IDEがあると便利です。

**Comment by Mnakagawa** [comment]
一つ忘れていました。
このテストはpython2.x系でしか動作しません。
python3.xではエラーが発生するため、その点ご了承ください。

ここまでの文章、READMEに書いといたほうがいいですね…

**Comment by cosmo0920** [comment]
おぉ、なるほど。ありがとうございます。

> ここまでの文章、READMEに書いといたほうがいいですね…

お願いします！

至る箇所にtypoっぽいのも散見されるので直して頂けると意味が取りやすくなって良いと思います。

```log
regist -> register
chenge -> change
```

etc. :sweat:

**Comment by Mnakagawa** [comment]
:cry:

**Comment by cosmo0920** [comment]
@Mnakagawa さん、泣かないでくださいー。
https://github.com/project-hatohol/web-gui-test はこのリポジトリに合流させてfeatureテストをTravisでも流せるようにする予定はありますか？
それとも引き続き上記リポジトリで更新を続ける予定ですか？

というのも、featureテストがあった方がモーダルダイアログをjQuery-uiからBootstrapのそれに置き換える時にどこか壊していないか不安になりつつ作業しなくても良くなるからです。

**Comment by Mnakagawa** [comment]
@cosmo0920 さん

Travisで流せるようにしたほうがいいとは思いますが、このseleniumを用いたテストは
不安定です。実際にブラウザを起動してページを開き、その中に指定したDOMがあるかを
チェックし、なければエラーが帰ってきてしまいます。

マシンの稼働率によってブラウザのパフォーマンスが変わりますが、その影響をもろに受けてしまうため
Travisで運用するレベルにまで達していないのです。時間が無いため中途半端なまま放置していますが、
本来であればきちんと運用出来るレベルまで持って行きたいところです。

もう一点問題がある可能性があって、このテストは実際にブラウザが立ち上がらない場合
（TravisはCUIで動作しているため、ブラウザを出力するディスプレイがない）に、seleniumのテストを行うことが出来ません。

仮想ディスプレイを用意すればよいのですが、Travisでそういったものを導入できますか？
Travisの知識が全く無いため、そのあたりよくわかりません。

**Comment by cosmo0920** [comment]
@Mnakagawa さん、できますよ。Travis CIでGUIのテストを行うには、
http://docs.travis-ci.com/user/gui-and-headless-browsers/#Starting-a-Web-Server
にあるように、.travis.ymlで

```yaml
before_script:
  - "export DISPLAY=:99.0"
  - "sh -e /etc/init.d/xvfb start"
```

としてやればGUI環境が中で立ち上がるようにできます。

> Travisで流せるようにしたほうがいいとは思いますが、このseleniumを用いたテストは
> 不安定です。実際にブラウザを起動してページを開き、その中に指定したDOMがあるかを
> チェックし、なければエラーが帰ってきてしまいます。

うーん、そうですかぁ。
一応こちらでも代案としてCasper.jsでfeatureテストを書いてみた https://github.com/project-hatohol/hatohol/compare/experimental/feature-test-casperjs のですが、要素が出現するまで待つマジックナンバーが現れてしまって苦戦しています。
こちらの方はPhantomJSで動くのでxvfbは必要ないです。

他にもWebのfeatureテストツールにはRubyを使えるならCapybaraなどもあります。
が、Capybara用にHTMLを手直ししないといけない箇所がありそうです。

なんにせよWebUIのテストは茨の道ということが分かりました…。 :cold_sweat:

**Comment by Mnakagawa** [comment]
> 要素が出現するまで待つマジックナンバーが現れてしまって苦戦しています。

恐らく同様の問題ですね:sob:
手動がそれなりに大変なので自動化したいのはやまやまなのですが、どうしたものか…

**Comment by cosmo0920** [comment]
Casper.jsを用いたWebUIのfeatureテストがTravis CIで回るようになりました。:clap:
https://travis-ci.org/project-hatohol/hatohol/jobs/51162684#L1808

**Comment by Mnakagawa** [comment]
おお〜〜〜〜！！すごい。おめでとうございます:sparkles:

**Comment by cosmo0920** [comment]
こちらはこちらでCasper.jsでのfeatureテストのテストケースを追加していこうと思います。
Seleniumのテストはどうしましょう？メンテナンス続けますか？

**Comment by Mnakagawa** [comment]
CasperでのテストがTravis上で動作するのであれば、そちらにテストケースを追加していったほうが
効率的だと思います（というよりはseleniumもメンテナンスしていくのは二度手間）。その場合は
seleniumのリポジトリも削除したほうがいいですね。

このあたりは @kz0817 さんに判断を仰いだほうが良いと思います。

**Comment by cosmo0920** [comment]
そうですね。Casper.jsを使うことにするのであれば現状のSeleniumのテストをCasper.jsで置き換えられた段階で削除するのが良いと思います。

ひとまず判断待ち了解です。

**Comment by cosmo0920** [comment]
メモ：今のmasterではWebUIのテストはmochaとCasper.jsの両方で可能になっている。しかしCasper.jsのテストはまだ不安定なこともあり、今の時点ではmocha側に追加していくようにしたい。
Casper.jsのテストは主にユーザーの入力をエミュレーションすることが目的。
